app/models/commit_status.rb

```diff
@@ -2,24 +2,23 @@
 # Used to display all warnings/failures before user actually deploys
 class CommitStatus
   # See ref_status_typeahead.js for how statuses are handled
-  STATUS_PRIORITY = {
-    success: 0,
-    pending: 1,
-    failure: 2,
-    error: 3,
-    fatal: 4
-  }.freeze
-
-  def initialize(stage, reference)
-    @stage = stage
+  # See https://developer.github.com/v3/repos/statuses for api details
+  # - fatal is our own state that blocks deploys
+  # - missing is our own state that means we could not determine the status
+  STATE_PRIORITY = [:success, :pending, :missing, :failure, :error, :fatal].freeze
+  UNDETERMINED = ["pending", "missing"].freeze
+
+  def initialize(project, reference, stage: nil)
+    @project = project
     @reference = reference
+    @stage = stage
   end
```

Comment:
If you're going to do this, why not make CommitStatus _less_ aware of our domain? I'd just give it a Git sha., since I don't think you need project/stage in the cache key, right?